### ISSUE TITLE
Adding a horizontal line before the footnote content

### ISSUE THREAD

**Issue by agmarrugo** [gettalong/kramdown#14]
I'd really like to separate the footnote content with a horizontal line. I've seen that  in

```
# File lib/kramdown/converter/html.rb, line 397
def footnote_content
```

The footnote content is generated as 

```
"<div class=\"footnotes\">\n#{convert(ol, 2)}</div>\n"
```

I think it would be better to add a `</hr>` so that the footnotes are separated from the main content.

**Comment by CarterPape** [comment]
@gettalong, I definitely agree with @agmarrugo that this should be done. [Daring Fireball](https://daringfireball.net/2018/05/duplex_booked_restaurant)'s CMS does this automatically somehow, and I think it's a good illustration of why: borders (especially on block elements) are limited in their role as dividers compared to `<hr />`s, which exist for the purpose of being what is asked for here: a logical, horizontal, customizable divider between the `.footnotes` element and the rest of the text. Asking developers to use CSS to define a border in this way is limiting.

I'd like to see this issue reopened, but I'd also like to hear a compelling reason not to if such is a reason exists.

**Comment by gettalong** [comment]
@CarterPape The `<hr />` serves no real purpose in the output except visually offsetting the footnotes.

If you really need to change the output to your tastes, just subclass the HTML converter. This would be the best way.

Otherwise see https://kramdown.gettalong.org/converter/html.html#footnotes for a way to set the location of the footnotes and use the [horizontal rule syntax](https://kramdown.gettalong.org/syntax.html#horizontal-rules) to insert the rule.

**Comment by agmarrugo** [comment]
@CarterPape I ended up doing it via CSS styling. I still have an octopress (Jekyll) blog that uses kramdown. I've been meaning to change to something else, but haven't had the time. What do you think?

```css
div.footnotes {
	font-size: 0.8em;
/*   padding-left: $pad-min;*/
	hr {
/*	padding-left: 0px;*/
    	border: none;
    	background-color: #ccc;
    	color: #ccc;
    	height: 1px;
	}
	ol {
		padding: 0px 5px 5px 30px;
	}
}
```

**Comment by letrastudio** [comment]
Hi, I'm hoping I might be able to offer a more compelling argument for this change.

> The `<hr />` serves no real purpose in the output except visually offsetting the footnotes.

I know it appears that way, but this isn't really true. Instead of a purely visual decoration, the horizontal rule element has semantic meaning. This change was introduced in HTML5:

> Although previous versions of HTML defined the `hr` element only in presentational terms, the element has now been given the specific semantic purpose of representing a “paragraph-level thematic break”.

Source: <http://w3c.github.io/html-reference/hr.html>.

> The `hr` element represents a paragraph-level thematic break, e.g. a scene change in a story, or a transition to another topic within a section of a reference book.

Source: <https://www.w3.org/TR/html50/grouping-content.html#the-hr-element>

This semantic definition seems appropriate for separating footnotes. But the benefits are not just theoretical. Here are two practical examples of how this would make a difference:

- if you're using a screen reader, an `<hr>` element will be described out loud, whereas a CSS border will not; adding a horizontal rule will make the document more accessible, allowing a screen reader user (or someone using other assistive technologies) to always notice some sort of break where the document body ends and the footnotes begin

- if you're using a CSS border and: a) styles fail to load; b) the user has styles turned off; or c) a "Reader" mode is activated (like the one in Safari), there will be no visual separation between content and footnotes; an `<hr>` guarantees that visual separation in every case

It's true that authors can simply add a horizontal rule to the bottom of the document themselves, but on that I think there is a bit of a philosophical issue at play: Markdown should be human-readable and understandable as-is, regardless of how HTML output is handled. [John Gruber](https://daringfireball.net/projects/markdown/):

> The overriding design goal for Markdown’s formatting syntax is to make it as readable as possible. The idea is that a Markdown-formatted document should be publishable as-is, as plain text, without looking like it’s been marked up with tags or formatting instructions.

Adding a single horizontal rule at the bottom of a document looks like a mistake if you're just reading the Markdown code — it's not understable outside the context of how it will be processed by this particular parser.

So I really think this should be added as a configuration option, perhaps even be set as the default in a future version when other breaking changes are introduced.

To sum up, here's what output currently looks like:

```html
<div class="footnotes">
  <ol>
    ...
  </ol>
</div>
```

Here's what's being proposed, as an option:

```html
<div class="footnotes">
  <hr />
  <ol>
    ...
  </ol>
</div>
```

And I might be going too far here, but maybe Kramdown should give the user enough configuration leeway to achieve output like this, which would be even more accessible and semantically correct:

```html
<section class="footnotes">
  <h2>Footnotes</h2>
  <ol>
    ...
  </ol>
</section>
```

Of course, just adding the `<hr>` would be much simpler, and a very nice improvement on its own.

I hope to have added some perspective. If you agree that this change is worth implementing, I could possibly submit a PR. Thanks for reading! ✌️

**Comment by cmcknight** [comment]
I use the following in my Markdown file:

<hr style="width: 10rem; height: 1px; border: none; background: #000;" />

It gives me the visual effect I'm after while supporting the thematic idea for screen readers...
